### PR TITLE
Remove all RenderState/TextureState Patches

### PR DESCRIPTION
**Pull request by LukeUsher**
With this work, all of these patches are no longer required, and instead, are replaced with a RenderStateConverter and TextureStateConverter object.

This completely removes a full 41 HLE patches!

As an aside, this also removes the 3925 fog hack, which is no longer required to get graphics in Halo or Silent Hill 2.

Note: this is a massive change to Cxbx-Rs rendering system, and as such, regressions are likely. Please report them here and I’ll look at them

### CODE REVIEW

**Comment by LukeUsher** [comment]
image
image
image

**Comment by gandalfthewhite19890404** [comment]
WWE Raw 2 is crashing:
Received Exception Code 0xC0000005 @ EIP := 0x042EECE8

**Comment by LukeUsher** [comment]
@gandalfthewhite19890404 don't just say 'some letters in menus are missing' we need to know what it was like before and after this PR, were letters missing before? Is it the same ones? etc.

Please don't give vague feedback, just single line comments are useless to work from.

File: src/core/hle/D3D8/XbPixelShader.cpp [hunk 1]
```diff
@@ -6043,15 +6043,13 @@ VOID XTL::DxbxUpdateActivePixelShader() // NOPATCH
...
     // needed for the final combiner constants at least)!
 
-/* This must be done once we somehow forward the vertex-shader oFog output to the pixel shader FOG input register :
-   We could use the unused oT4.x to output fog from the vertex shader, and read it with 'texcoord t4' in pixel shader!
-    // Disable native fog if pixel shader is said to handle it :
-    if ((RecompiledPixelShader.PSDef.PSFinalCombinerInputsABCD > 0)
-    || (RecompiledPixelShader.PSDef.PSFinalCombinerInputsEFG > 0))
-    {
-      g_pD3DDevice->SetRenderState(D3DRS_FOGENABLE, BOOL_FALSE);
+    // TODO: Figure out a method to foward the vertex-shader oFog output to the pixel shader FOG input register :
```
Comment (Margen67) [inline src/core/hle/D3D8/XbPixelShader.cpp]: *forward

**Comment by MojoJojoDojo** [comment]
image
There are missing graphics/textures on the character models here.
It worked fine beforehand.
GPU : AMD Radeon 530
Game : Dead Or Alive.
I didnt use any wrappers.

**Comment by Margen67** [comment]
The Bible Game no longer renders a black screen:
image

File: src/core/hle/D3D8/XbConvert.cpp [hunk 2]
```diff
@@ -1418,16 +1418,16 @@ const RenderStateInfo DxbxRenderStateInfo[] = {
...
-	{ "D3DRS_STIPPLEENABLE"               /*=  83*/, 4627, xtBOOL,                NV2A_POLYGON_STIPPLE_ENABLE },
-	{ "D3DRS_SIMPLE_UNUSED8"              /*=  84*/, 4627, xtDWORD,               0 },
-	{ "D3DRS_SIMPLE_UNUSED7"              /*=  85*/, 4627, xtDWORD,               0 },
-	{ "D3DRS_SIMPLE_UNUSED6"              /*=  86*/, 4627, xtDWORD,               0 },
-	{ "D3DRS_SIMPLE_UNUSED5"              /*=  87*/, 4627, xtDWORD,               0 },
-	{ "D3DRS_SIMPLE_UNUSED4"              /*=  88*/, 4627, xtDWORD,               0 },
-	{ "D3DRS_SIMPLE_UNUSED3"              /*=  89*/, 4627, xtDWORD,               0 },
-	{ "D3DRS_SIMPLE_UNUSED2"              /*=  90*/, 4627, xtDWORD,               0 },
-	{ "D3DRS_SIMPLE_UNUSED1"              /*=  91*/, 4627, xtDWORD,               0 },
+	{ "D3DRS_DEPTHCLIPCONTROL"            /*=  82*/, 4531, xtD3DDCC,              NV2A_DEPTHCLIPCONTROL },
```
Comment (LukeUsher) [inline src/core/hle/D3D8/XbConvert.cpp]: These changes may not be 100% correct, needs to be tested and verified with as many games as possible.

**Comment by gandalfthewhite19890404** [comment]
Result with 35393e6e

---------------------------------------------------------------------------------------------------------------
- Euro 2004
<details><summary>Regress if compare to develop branch. But you can see booting and loading progress (green squares instead of black screen)</summary>

image
</details>

---------------------------------------------------------------------------------------------------------------

File: src/core/hle/D3D8/XbConvert.cpp (hunk 2 above)
Comment (LukeUsher) [reply src/core/hle/D3D8/XbConvert.cpp]: They're not, they're going to be reverted

**Comment by LukeUsher** [comment]
Many regressions have been ironed out with the latest commit, so games should be retested if possible

**Comment by gandalfthewhite19890404** [comment]
Sure, I will

File: CMakeLists.txt [hunk 3]
```diff
@@ -250,14 +250,15 @@ file (GLOB CXBXR_SOURCE_EMU
...
+ "${CXBXR_ROOT_DIR}/src/core/hle/D3D8/Direct3D9/TextureStates.cpp"
  "${CXBXR_ROOT_DIR}/src/core/hle/D3D8/Direct3D9/WalkIndexBuffer.cpp"
  "${CXBXR_ROOT_DIR}/src/core/hle/D3D8/ResourceTracker.cpp"
  "${CXBXR_ROOT_DIR}/src/core/hle/D3D8/XbConvert.cpp"
  "${CXBXR_ROOT_DIR}/src/core/hle/D3D8/XbD3D8Logging.cpp"
  "${CXBXR_ROOT_DIR}/src/core/hle/D3D8/XbPixelShader.cpp"
  "${CXBXR_ROOT_DIR}/src/core/hle/D3D8/XbPushBuffer.cpp"
- "${CXBXR_ROOT_DIR}/src/core/hle/D3D8/XbState.cpp"
- "${CXBXR_ROOT_DIR}/src/core/hle/D3D8/XbVertexBuffer.cpp"
+  "${CXBXR_ROOT_DIR}/src/core/hle/D3D8/XbVertexBuffer.cpp"
```
Comment (RadWolfie) [inline CMakeLists.txt]: Remove the extra space here, thanks.

File: src/core/hle/D3D8/Direct3D9/RenderStates.h [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+// *
+// *  This program is distributed in the hope that it will be useful,
+// *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+// *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// *  GNU General Public License for more details.
+// *
+// *  You should have recieved a copy of the GNU General Public License
+// *  along with this program; see the file COPYING.
+// *  If not, write to the Free Software Foundation, Inc.,
+// *  59 Temple Place - Suite 330, Bostom, MA 02111-1307, USA.
```
Comment (RadWolfie) [inline src/core/hle/D3D8/Direct3D9/RenderStates.h]: Boston*

File: src/core/hle/D3D8/Direct3D9/TextureStates.cpp [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+// *
+// *  This program is distributed in the hope that it will be useful,
+// *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+// *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// *  GNU General Public License for more details.
+// *
+// *  You should have recieved a copy of the GNU General Public License
+// *  along with this program; see the file COPYING.
+// *  If not, write to the Free Software Foundation, Inc.,
+// *  59 Temple Place - Suite 330, Bostom, MA 02111-1307, USA.
```
Comment (RadWolfie) [inline src/core/hle/D3D8/Direct3D9/TextureStates.cpp]: Same here

File: src/core/hle/D3D8/Direct3D9/RenderStates.h (hunk 4 above)
Comment (LukeUsher) [reply src/core/hle/D3D8/Direct3D9/RenderStates.h]: Job for another PR, it's typed this way in almost every source file, in both Cxbx-R and XbSymbolDatabase!

File: src/core/hle/D3D8/Direct3D9/TextureStates.cpp (hunk 5 above)
Comment (LukeUsher) [reply src/core/hle/D3D8/Direct3D9/TextureStates.cpp]: See previous reply, it's in every code file, not just these new ones.

**Comment by gandalfthewhite19890404** [comment]
47553c63
Morrowind - water is still bad.
LEGO Star Wars The Original Trilogy - fixed, now looks like in develop.
LOTR Third Age - same as earlier in PR1686.

**Comment by ESPPIRAL** [comment]
Games that now render with "rainbow" textures.

Vexx

bandicam 2019-07-30 19-19-14-391
BuildVersion = 5120

Splinter Cell Double Agent.

bandicam 2019-07-30 21-48-29-860
bandicam 2019-07-30 21-48-36-808

BuildVersion = 5849

Splinter Cell Chaos Theory.

bandicam 2019-07-30 21-53-05-941

BuildVersion = 5849

**Comment by ESPPIRAL** [comment]
Spiderman

bandicam 2019-07-30 22-03-59-449
BuildVersion = 4134

**Comment by RadWolfie** [comment]
Tested as of commit 2d6999d

Regressed:
* <details><summary>Turok Evolution</summary>

   Turok_pr_menu
   Turok_pr_ingame
   </details>
* GTA 3 and GTA: Vice City
  * If use a fresh installed, both titles doesn't work correctly. Except GTA: Vice City will perform a reboot to display an error from a dashboard (if installed).

Improved:
*  <details><summary>Halo Demo</summary>

   Halo_ingame_part1
   Halo_ingame_part2
   Halo_ingame_part3
   </details>

**Comment by ESPPIRAL** [comment]
Shreck 2 

bandicam 2019-07-30 22-10-36-444

BuildVersion = 5788

**Comment by ESPPIRAL** [comment]
Rally Sport Challenge the car is a white box, previously it was rendering correctly.

bandicam 2019-07-30 22-18-34-428

BuildVersion = 4134

**Comment by LukeUsher** [comment]
Okay that's enough :P
lots of regressions that I didn't expect, I'll reopen when they're solved.
They all appear to be similar issues.

**Pull request closed without merging**